File: backend/src/models/Class.js
const pool = require('../config/database');

const Class = {
  async findAll() {
    const { rows } = await pool.query('SELECT * FROM classes ORDER BY label');
    return rows;
  },

  async findById(id) {
    const { rows } = await pool.query('SELECT * FROM classes WHERE id = $1', [id]);
    return rows[0] || null;
  },

  async create({ label, year }) {
    const { rows } = await pool.query(
      'INSERT INTO classes (label, year) VALUES ($1, $2) RETURNING *',
      [label, year || null]
    );
    return rows[0];
  },

  async update(id, fields) {
    const updates = [];
    const values = [];

    if (fields.label !== undefined) {
      values.push(fields.label);
      updates.push(`label = $${values.length}`);
    }
    if (fields.year !== undefined) {
      values.push(fields.year);
      updates.push(`year = $${values.length}`);
    }

    if (updates.length === 0) return null;

    values.push(id);
    const { rows } = await pool.query(
      `UPDATE classes SET ${updates.join(', ')} WHERE id = $${values.length} RETURNING *`,
      values
    );
    return rows[0] || null;
  },

  async delete(id) {
    const { rowCount } = await pool.query('DELETE FROM classes WHERE id = $1', [id]);
    return rowCount > 0;
  },
};

module.exports = Class;

File: backend/src/models/class_.py
from ..config.database import query, query_one


class ClassModel:
    @staticmethod
    def find_all() -> list[dict]:
        return query("SELECT * FROM classes ORDER BY label")

    @staticmethod
    def find_by_id(id: int) -> dict | None:
        return query_one("SELECT * FROM classes WHERE id = %s", (id,))

    @staticmethod
    def create(label: str, year: int | None = None) -> dict:
        return query_one(
            "INSERT INTO classes (label, year) VALUES (%s, %s) RETURNING *",
            (label, year),
        )

    @staticmethod
    def update(id: int, fields: dict) -> dict | None:
        updates, values = [], []
        if "label" in fields:
            updates.append("label = %s")
            values.append(fields["label"])
        if "year" in fields:
            updates.append("year = %s")
            values.append(fields["year"])
        if not updates:
            return None
        values.append(id)
        return query_one(
            f"UPDATE classes SET {', '.join(updates)} WHERE id = %s RETURNING *",
            values,
        )

    @staticmethod
    def delete(id: int) -> bool:
        rows = query("DELETE FROM classes WHERE id = %s RETURNING id", (id,))
        return len(rows) > 0
